plugins/virtual_folders/server/__init__.py

```diff
@@ -0,0 +1,125 @@
+        if list(Folder().find({
+            'parentId': doc['_id'],
+            'parentCollection': 'folder'
+        }, limit=1)):
+            raise ValidationException(
+                'Virtual folders may not contain child folders.', field='isVirtual')
+    if doc['parentCollection'] == 'folder':
+        parent = Folder().load(event.info['parentId'], force=True, exc=True)
+        if parent.get('isVirtual'):
+            raise ValidationException(
+                'You may not place folders under a virtual folder.', field='folderId')
+
+    if 'virtualItemsQuery' in doc:
+        try:
+            json.loads(doc['virtualItemsQuery'])
+        except (TypeError, ValueError):
+            raise ValidationException(
+                'The virtual items query must be valid JSON.', field='virtualItemsQuery')
+
+    if 'virtualItemsSort' in doc:
+        try:
+            json.loads(doc['virtualItemsSort'])
+        except (TypeError, ValueError):
+            raise ValidationException(
+                'The virtual items sort must be valid JSON.', field='virtualItemsSort')
+
+
+def _validateItem(event):
+    parent = Folder().load(event.info['folderId'], force=True, exc=True)
+    if parent.get('isVirtual'):
+        raise ValidationException(
+            'You may not place items under a virtual folder.', field='folderId')
+
+
+@rest.boundHandler
+def _folderUpdate(self, event):
+    params = event.info['params']
+    if {'isVirtual', 'virtualItemsQuery', 'virtualItemsSort'} & set(params):
+        folder = Folder().load(event.info['returnVal']['_id'], force=True)
+        update = False
+
+        if params.get('isVirtual') is not None:
+            update = True
+            folder['isVirtual'] = params['isVirtual']
+        if params.get('virtualItemsQuery') is not None:
+            update = True
+            folder['virtualItemsQuery'] = params['virtualItemsQuery']
+        if params.get('virtualItemsSort') is not None:
+            update = True
+            folder['virtualItemsSort'] = params['virtualItemsSort']
+
+        if update:
+            self.requireAdmin(self.getCurrentUser(), 'Must be admin to setup virtual folders.')
+            folder = Folder().filter(Folder().save(folder), rest.getCurrentUser())
+            event.preventDefault().addResponse(folder)
+
+
+@access.public
+@rest.boundHandler
+def _virtualChildItems(self, event):
+    params = event.info['params']
+
+    if 'folderId' not in params:
+        return  # This is not a child listing request
+
+    user = self.getCurrentUser()
+    folder = Folder().load(params['folderId'], user=user, level=AccessType.READ)
+
+    if not folder.get('isVirtual') or 'virtualItemsQuery' not in folder:
+        return  # Parent is not a virtual folder, proceed as normal
+
+    limit, offset, sort = self.getPagingParameters(params, defaultSortField='name')
+    q = json.loads(folder['virtualItemsQuery'], object_hook=json_util.object_hook)
```

Comment:
Should we be using `bson.json_util.loads` rather than passing `json_util.object_hook` to `json.loads`?  It uses `object_hook_pairs` rather than `object_hook` (at least in the versions of Python we care about).